### PR TITLE
Refactoring of Database assembly

### PR DESCRIPTION
**Pull request by Eastrall**
This PR is a refactoring of the `Rhisis.Database` assembly and solves issue #3 and #4.
It includes the following:
- New `DatabaseFactory` class as Singleton to create a database context
- Repositories can work with or without db context. (If no context is assigned, the repository will create a new one)
- Add unit tests for `DatabaseFactory`
- Add unit tests for `UserRepository` (Need to make more for other repositories)
- Add a new database migration script : `CreateMigration.ps1`

---

The difference between the old version of the database access layer and this one resides mostly on the repository and database factory.
In the previous version, each time you want to make a database request, you would need to declare the following statements:
```csharp
using (var db = DatabaseService.GetContext())
{
     var user = db.Users.Get(x => x.Id == id);
     var character db.Characters.Get(...);

     // And so on...
}
```
Now, you have two possibilities:

### Declare a new database context using the `DatabaseFactory` class

```cs
using (var db = DatabaseFactory.Instance.CreateDbContext())
{
    var userRepository = new UserRepository(db);
    var characterRepository = new CharacterRepository(db);

    // Do work here with repositories
}
```
This allows the developer to has a single context and use multiples repositories that shares this context.
Useful when entities are being tracked by EFCore. 😉 

### Declare the repository you want to use

```csharp
var userRepository = new UserRepository();
DbUser user = userRepository.Get(x => x.Username == "Eastrall");
```

This way, everytime you will do a request, it will create a new context and make the request.

### CODE REVIEW

File: src/Rhisis.Database/DatabaseContext.cs [hunk 1]
```diff
@@ -1,139 +1,108 @@
...
-using Rhisis.Database.Interfaces;
-using Rhisis.Database.Repositories;
 using System;
-using System.IO;
+using System.Threading.Tasks;
 
 namespace Rhisis.Database
 {
-    public class DatabaseContext : DbContext
+    public sealed class DatabaseContext : DbContext
```
Comment (Skeatwin) [inline src/Rhisis.Database/DatabaseContext.cs]: Yes !

Comment (Eastrall) [reply src/Rhisis.Database/DatabaseContext.cs]: We prevent this class from being inherited from others. Creating of a new `DatabaseContext` should be done by using the `DatabaseFactory.CreateDbContext()` method.

File: src/Rhisis.Database/DatabaseContext.cs [hunk 2]
```diff
@@ -1,139 +1,108 @@
...
     {
         private const string MySqlConnectionString = "server={0};userid={1};pwd={2};port={4};database={3};sslmode=none;";
         private const string MsSqlConnectionString = "Server={0},{1};Database={2};User Id={3};Password={4};";
-        private static readonly string DefaultConfigFile = Path.Combine(Environment.CurrentDirectory, "..", "..", "bin", "config", "database.json");
-        private readonly DatabaseConfiguration _configuration;
+
+        private readonly DatabaseConfiguration _databaseConfiguration;
```
[the remark this answers is not in the document]
Comment (Eastrall) [reply src/Rhisis.Database/DatabaseContext.cs]: I totally agree with you. I'll change this.

File: test/Rhisis.Database.Test/DatabaseContextTest.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
+﻿using Microsoft.EntityFrameworkCore;
+using System.Linq;
+using Xunit;
+
+namespace Rhisis.Database.Test
+{
+    public class DatabaseContextTest
+    {
+        [Fact(DisplayName = "Create InMemory context")]
```
[the remark this answers is not in the document]
Comment (Eastrall) [reply test/Rhisis.Database.Test/DatabaseContextTest.cs]: True. I'll change this. 😉

[the remark this answers is not in the document]
Comment (Skeatwin) [reply test/Rhisis.Database.Test/DatabaseContextTest.cs]: Only visual but we are so exigents :laughing:

**Pull request merged**